text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,774 @@
+This is due to the high implementation effort, see below, this restriction can be lifted by a future RFC.
+
+Closures use the `rust-call` unstable calling convention, which would need to be adapted to guarantee TCE.
+Additionally, any closure that has captures would need special handling, since the captures would currently be dropped before the tail call.
+
+## Variadic functions using `c_variadic`
+
+Tail calling [variadic functions](https://doc.rust-lang.org/beta/unstable-book/language-features/c-variadic.html) _and_ tail calling _from_ variadic functions is **not** allowed.
+As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but supported can be added with a future RFC.
+
+## Generators
+
+Tail calling from [generators](https://doc.rust-lang.org/beta/unstable-book/language-features/generators.html) is **not** allowed. As the generator state is stored internally, tail calling from the generator function would require additional support to function correctly. To limit implementation effort this is not supported but can be supported by a future RFC.
+
+## Async
+[async]: #async
+
+Tail calling _from_ async functions is **not** allowed, neither calling async nor calling sync functions is supported. This is due to the high implementation effort as it requires special handling for the async state machine. This restriction can be relaxed by a future RFC.
+
+Using `become` on a `.await` call, such as `become f().await`, is also **not** allowed. This is because when using `.await`, the `Future` returned by `f()` is not "called" but run by the executor, thus, tail calls do not apply here.
+
+Note that tail calling async functions from sync code is possible but the return type for async functions is `std::future::Future`, which is unlikely to be interesting.
+
+## Operators are not supported
+
+Invocations of operators were considered as valid targets but were rejected on grounds of being too error-prone.
+In any case, these can still be called as methods. One example of their error-prone nature ([source](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1167112296)):
+```rust
+pub fn fibonacci(n: u64) -> u64 {
+    if n < 2 {
+        return n
+    }
+    become fibonacci(n - 2) + fibonacci(n - 1)
+}
+```
+In this case, a naive author might assume that this is going to be a stack space-efficient implementation since it uses tail recursion instead of normal recursion. However, the outcome is more or less the same since the critical recursive calls are not actually in tail call position.
+
+Further confusion could result from the same-signature restriction where the Rust compiler raises an error since fibonacci and <u64 as Add>::add do not share a common signature.
```

Comment:
```suggestion
Further confusion could result from the same-signature restriction where the Rust compiler raises an error since fibonacci and `<u64 as Add>::add` do not share a common signature.
```